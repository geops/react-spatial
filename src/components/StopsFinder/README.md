#

This demonstrates the use of the StopsFinder component.

```jsx
import React from 'react';
import { Layer } from 'mobility-toolbox-js/ol';
import Map from 'ol/Map';
import Tile from 'ol/layer/Tile';
import { fromLonLat } from 'ol/proj';
import OSM from 'ol/source/OSM';
import BasicMap from 'react-spatial/components/BasicMap';
import StopsFinder from 'react-spatial/components/StopsFinder';

const map = new Map({ controls: [] });

const layers = [
  new Layer({
    olLayer: new Tile({
      source: new OSM(),
    }),
  }),
];

// The `apiKey` used here is for demonstration purposes only.
// Please get your own api key at https://developer.geops.io/.
const { apiKey } = window;

<>
  <BasicMap
    map={map}
    center={[951560, 6002550]}
    zoom={14}
    layers={layers}
    tabIndex={0}
  />

  <StopsFinder
    map={map}
    apiKey={apiKey}
    onSelect={({ geometry }) => {
      map.getView().setCenter(fromLonLat(geometry.coordinates));
    }}
    autocompleteProps={{
      textFieldProps: {
        label: 'Search for stops',
      },
    }}
  />
</>
```
