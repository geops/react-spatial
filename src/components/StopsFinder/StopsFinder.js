import React, { useMemo, useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Autocomplete } from '@material-ui/lab';
import { FaSearch } from 'react-icons/fa';
import TextField from '@material-ui/core/TextField';
import CircularProgress from '@material-ui/core/CircularProgress';
import StopsFinderControl from 'mobility-toolbox-js/ol/controls/StopsFinderControl';
import { Map } from 'ol';
import { makeStyles } from '@material-ui/core';
import StopsFinderOptions from './StopsFinderOption';

const useStyles = makeStyles(() => {
  return {
    popupIndicatorOpen: {
      transform: 'rotate(0)',
    },
  };
});

function StopsFinder({
  agencies,
  apiKey,
  autocompleteProps,
  bbox,
  field,
  limit,
  map,
  mots,
  onSelect,
  radius,
  refLocation,
  renderAutocomplete,
  url,
}) {
  const classes = useStyles();
  const [inputValue, setInputValue] = useState('');
  const [suggestions, setSuggestions] = useState([]);
  const [isLoading, setLoading] = useState(false);
  const [isOpen, setOpen] = useState(false);

  const control = useMemo(() => {
    return new StopsFinderControl({
      url,
      apiKey,
      target: document.createElement('div'),
      element: document.createElement('div'),
      render(newSuggestions = []) {
        setSuggestions(newSuggestions);
        setLoading(false);
      },
    });
  }, [apiKey, url]);

  useEffect(() => {
    if (!inputValue) {
      setSuggestions([]);
      setLoading(false);
      return () => {};
    }
    const abortController = new AbortController();
    setLoading(true);
    control.apiParams = {
      prefAgencies: agencies && agencies.toString(),
      bbox: bbox && bbox.toString(),
      field: field && field.toString(),
      limit,
      mots: mots && mots.toString(),
      radius,
      ref_location: refLocation && refLocation.toString(),
    };
    control.search(inputValue, abortController);
    return () => {
      abortController.abort();
    };
  }, [
    agencies,
    bbox,
    control,
    field,
    inputValue,
    limit,
    mots,
    radius,
    refLocation,
  ]);

  // Ensure the control is not associated to the wrong map
  useEffect(() => {
    if (!control) {
      return () => {};
    }

    control.map = map;

    return () => {
      control.map = null;
    };
  }, [map, control]);

  if (!control) {
    return null;
  }

  if (renderAutocomplete) {
    return renderAutocomplete(
      suggestions,
      inputValue,
      setInputValue,
      isOpen,
      setOpen,
      isLoading,
      setLoading,
    );
  }

  return (
    <Autocomplete
      fullWidth
      autoComplete
      autoHighlight
      selectOnFocus
      getOptionLabel={(option) => option.properties.name}
      onChange={(evt, value, reason) => {
        if (onSelect && reason === 'select-option') {
          onSelect(value, evt);
        }
      }}
      popupIcon={<FaSearch focusable={false} size={15} />}
      renderInput={(params) => {
        return (
          <TextField
            label="Search stops"
            // eslint-disable-next-line react/jsx-props-no-spreading
            {...{
              ...params,
              ...((autocompleteProps || {}).textFieldProps || {}),
            }}
            InputProps={{
              ...params.InputProps,
              endAdornment: (
                <>
                  {isLoading && <CircularProgress size={20} />}
                  {params.InputProps.endAdornment}
                </>
              ),
            }}
          />
        );
      }}
      renderOption={(option) => {
        return <StopsFinderOptions option={option} />;
      }}
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...{ ...autocompleteProps, textFieldProps: null }}
      classes={{ ...classes, ...autocompleteProps.classes }}
      inputValue={inputValue}
      open={isOpen}
      options={suggestions}
      loading={isLoading}
      onOpen={() => {
        setOpen(true);
      }}
      onClose={() => {
        setOpen(false);
      }}
      onInputChange={(evt, val) => {
        setInputValue(val);
      }}
    />
  );
}

StopsFinder.propTypes = {
  /**
   * Array or a comma separated list of agencies which should be available.
   * Order of these agencies chooses which agency will be preferred.
   * Available values : sbb, db
   */
  agencies: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),

  /**
   * geOps api key to access the StopsFinder service.
   */
  apiKey: PropTypes.string,

  /**
   * Properties apply to the default [MUI Autocomplete component](https://material-ui.com/api/autocomplete/).
   * We add a custom properties textFieldProps for the default [MUI TextField component](https://material-ui.com/api/text-field/) used by the Autocomplete.
   */
  autocompleteProps: PropTypes.object,

  /**
   * minX,minY,maxX,maxY coordinates in WGS84 wherein the station should lie.
   */
  bbox: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.number),
  ]),

  /**
   * Array or a comma separated list of fields which should be used for look up.
   * Available values : id, name, coords
   */
  field: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),

  /**
   * Control how many matches will be returned.
   */
  limit: PropTypes.number,

  /**
   * A map.
   */
  map: PropTypes.instanceOf(Map).isRequired,

  /**
   * Array or a comma separated list of mode of transpaorts which should be available.
   * Available values : bus, ferry, gondola, tram, rail, funicular, cable_car, subway
   */
  mots: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]),

  /**
   * Function called when a suggestion is selected.
   */
  onSelect: PropTypes.func,

  /**
   * Radius around refLocation in meters that is most relevant.
   * Used as granularity for location rank.
   */
  radius: PropTypes.number,

  /**
   * Coordinates in WGS84 (in lat,lon order) used to rank stops close to this position higher.
   * Available values : id, name, coords
   */
  refLocation: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.number),
  ]),

  /**
   * Function to render a different autocomplete input than the default one.
   */
  renderAutocomplete: PropTypes.func,

  /**
   * Url of the geOps StopsFinder service.
   */
  url: PropTypes.string,
};

StopsFinder.defaultProps = {
  agencies: null,
  apiKey: null,
  autocompleteProps: {},
  bbox: null,
  field: null,
  limit: null,
  mots: null,
  onSelect: null,
  radius: null,
  refLocation: null,
  url: null,
  renderAutocomplete: null,
};

export default StopsFinder;
