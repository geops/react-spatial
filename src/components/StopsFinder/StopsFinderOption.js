import React, { lazy, Suspense } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core';

const ext = '_round-blue-01.svg';
const iconForMot = {};
[
  'bus',
  'ferry',
  'gondola',
  'tram',
  'rail',
  'funicular',
  'cable_car',
  'subway',
].forEach((mot) => {
  iconForMot[mot] = lazy(() => import(`../../images/mots/${mot}${ext}`));
});

const useStyles = makeStyles((theme) => {
  return {
    flex: {
      display: 'flex',
      alignItems: 'center',
    },
    icon: {
      marginRight: theme.spacing(2),
    },
  };
});

function StopsFinderOption({ option }) {
  const classes = useStyles();

  return (
    <Suspense fallback={<div />}>
      <div className={classes.flex}>
        {Object.entries(option.properties.mot).map(([key, value]) => {
          if (value) {
            const MotIcon = iconForMot[key];
            return (
              <span className={classes.icon}>
                <MotIcon />
              </span>
            );
          }
          return <></>;
        })}
        <span>{option.properties.name}</span>
      </div>
    </Suspense>
  );
}

StopsFinderOption.propTypes = {
  option: PropTypes.object.isRequired,
};

export default React.memo(StopsFinderOption);
