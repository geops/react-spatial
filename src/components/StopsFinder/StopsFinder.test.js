import React from 'react';
import renderer from 'react-test-renderer';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { Map } from 'ol';
import StopsFinder from '.';

configure({ adapter: new Adapter() });

describe('StopsFinder', () => {
  let map;

  beforeEach(() => {
    map = new Map({});
  });

  test('matches snapshots.', () => {
    const component = renderer.create(<StopsFinder map={map} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
