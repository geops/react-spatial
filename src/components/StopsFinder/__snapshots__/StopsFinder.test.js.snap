// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopsFinder matches snapshots. 1`] = `
<div
  aria-expanded={false}
  aria-owns={null}
  className="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  role="combobox"
  textFieldProps={null}
>
  <div
    className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
      id="undefined-label"
    >
      Search stops
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
      onClick={[Function]}
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="both"
        aria-controls={null}
        aria-invalid={false}
        autoCapitalize="none"
        autoComplete="off"
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedEnd"
        disabled={false}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        required={false}
        spellCheck="false"
        type="text"
        value=""
      />
      <div
        className="MuiAutocomplete-endAdornment"
      >
        <button
          aria-label="Clear"
          className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Clear"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Open"
          className="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          title="Open"
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              fill="currentColor"
              focusable={false}
              height={15}
              size={15}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width={15}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
